File: wordBreak.cpp
#include <bits/stdc++.h>
using namespace std;

/*
#include <string>
#include <sstream>
#include <vector>
#include <iterator>
*/
template<typename Out>
void split(const std::string &s, char delim, Out result) {
    std::stringstream ss(s);
    std::string item;
    while (std::getline(ss, item, delim)) {
        *(result++) = item;
    }
}

std::vector<std::string> split(const std::string &s, char delim) {
    std::vector<std::string> elems;
    split(s, delim, std::back_inserter(elems));
    return elems;
}








unordered_set<string> myDict;


void solutions(string sent, string outputD="") {
    //cout << outputD << endl;
/*
    vector<string> wordsTillNow = split(outputD, ' ');

    //wordsTillNow = outputD.split()
  	if (wordsTillNow.size() > 2) {
    	  for (int i = 0; i < wordsTillNow.size() - 1; i++)  // ULL
        {
          if (myDict.find(wordsTillNow[i]) == myDict.end())
            return;

        }
    }
*/

  /*
    if len(wordsTillNow) > 2:
        wordsDecided = wordsTillNow[:-1]
        for word in wordsDecided:
            if word not in setWords:
                return
                */

    if (sent.length() <= 1)
    {
        string completeSent = outputD + sent[0];
        vector<string> words = split(completeSent, ' ');
    	  for (int i = 0; i < words.size(); i++)  // ULL
        {
          if (myDict.find(words[i]) == myDict.end())
            return;
        }
        cout << (completeSent) << endl;
        return;
    }

    // decide NOT to split it
    solutions(sent.substr(1), outputD + sent[0]);
    // decode to split it
    solutions(sent.substr(1), outputD + sent[0] + " ");

}

int main(void) {
  int n;
  cin >> n;

  for (int i = 0; i < n; i++) {
    string s;
    cin >> s;
    myDict.insert(s);
  }
  string empty_str = "";
  string one_space = " ";
  myDict.insert(empty_str);
  myDict.insert(one_space);

  string sentence;
  cin >> sentence;
  solutions(sentence);
}

File: wordBreak.py



# iloveicecreamandmango
# i loveicecreamandmango,

# abc
# a
def solutions(sent, setWords, wordsTillNow):

    if len(wordsTillNow) > 2:
        for i in range(len(wordsTillNow) - 1):
            word = wordsTillNow[i]
            if word not in setWords:
                return
    if (len(sent) <= 1):
        wordsTillNow[-1] = wordsTillNow[-1] + sent[0]
        for word in wordsTillNow:
            if word not in setWords:
                return
        #print (completeSent)
        for word in wordsTillNow:
            print(word, end=' ')
        print()
        return

    solutions(sent[1:], setWords, wordsTillNow + [sent[0]])
    # deide NOTto split it
    wordsTillNow[-1] = wordsTillNow[-1] + sent[0]
    solutions(sent[1:], setWords, wordsTillNow)

def solutions2(sent, setWords, outputD=""):
    #break it down
    wordsTillNow = outputD.split()
    if len(wordsTillNow) > 2:
        wordsDecided = wordsTillNow[:-1]
        for word in wordsDecided:
            if word not in setWords:
                return

    if (len(sent) <= 1):
        completeSent = outputD + sent[0]
        words = completeSent.strip().split()
        for word in words:
            if word not in setWords:
                return
        print(completeSent)
        return
    # decode NOT to split it
    solutions(sent[1:], setWords, outputD + sent[0])
    # decode to split it
    solutions(sent[1:], setWords, outputD + sent[0] + " ")


#f = open("test.txt")
n = int(input()) # replace f.readline() with input()
myDict = set(input().strip() for i in range(n))
sentence = input()#"iloveicecreamandmango"#  f.readline()
#print(myDict)
# given a sentence and setOfWordsinMyDict figure out how many ways to splice a string in meangful way
puttingFirstThingIn = [sentence[0]]
solutions(sentence[1:], myDict, puttingFirstThingIn)
